File: src/main/java/com/pyamsoft/pydroid/app/service/AutoRestartService.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.app.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.SystemClock;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

public abstract class AutoRestartService extends Service {

  private AlarmManager alarmManager;

  @Override public void onCreate() {
    super.onCreate();
    alarmManager = (AlarmManager) getApplicationContext().getSystemService(Context.ALARM_SERVICE);
  }

  // Allow for restarting the service on KitKat when the task is removed. There is a bug that
  // otherwise prevents this from working. Ugly hack workaround should fix most cases.
  @Override public void onTaskRemoved(final @NonNull Intent rootIntent) {
    super.onTaskRemoved(rootIntent);
    final PendingIntent pendingIntent = PendingIntent.getService(getApplicationContext(), 1,
        new Intent(getApplicationContext(), getServiceClass()), PendingIntent.FLAG_UPDATE_CURRENT);
    alarmManager.cancel(pendingIntent);
    alarmManager.set(AlarmManager.ELAPSED_REALTIME, SystemClock.elapsedRealtime() + 4000,
        pendingIntent);
  }

  @NonNull @CheckResult protected abstract Class<? extends AutoRestartService> getServiceClass();
}

File: pydroid-ui/src/main/java/com/pyamsoft/pydroid/ui/app/AutoRestartService.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.pydroid.ui.app

import android.app.AlarmManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.SystemClock
import timber.log.Timber
import java.util.concurrent.TimeUnit

/**
 * Android API 19 has a bug where a service, even a foreground one, will stop when swiped away or
 * task
 * killed, even if START_STICKY is set
 *
 * This service is configured to auto restart itself when it is destroyed via onTaskRemoved, and
 * will
 * only reliably work with Foreground services. Background services may also work, but since these
 * are
 * generally limited in Android O, you should work to move away from them as best as possible.
 */
abstract class AutoRestartService : Service() {

  /**
   * Change to disable auto restart ability
   */
  protected open val isAutoRestartEnabled: Boolean = true

  override fun onTaskRemoved(rootIntent: Intent) {
    super.onTaskRemoved(rootIntent)
    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.KITKAT && isAutoRestartEnabled) {
      Timber.w("""
        |Android KitKat is affected by a bug which causes START_STICKY
        |and Foreground services to be killed onTaskRemoved.
      """.trimMargin())
      Timber.w(
          "The service will schedule an Alarm for 5 seconds out to automatically restart itself")

      val restartIntent = Intent(applicationContext, javaClass)
      restartIntent.`package` = packageName

      // We must use a non-zero value for the RC, Samsung seems to like things over 1000
      val pendingIntent = PendingIntent.getService(applicationContext, 1024, restartIntent,
          PendingIntent.FLAG_ONE_SHOT)

      val alarmManager = applicationContext.getSystemService(Context.ALARM_SERVICE) as AlarmManager

      // Cancel any old alarms
      alarmManager.cancel(pendingIntent)
      pendingIntent.cancel()

      // Schedule alarm
      val restartTime = SystemClock.elapsedRealtime() + TimeUnit.SECONDS.toMillis(5)
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        alarmManager.setAndAllowWhileIdle(AlarmManager.ELAPSED_REALTIME, restartTime, pendingIntent)
      } else {
        alarmManager.set(AlarmManager.ELAPSED_REALTIME, restartTime, pendingIntent)
      }
    }
  }
}
